source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+{
+    if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_WAITING )
+    {
+        if( xAzureIoTProvisioningClientHandle->_internal.xGetTimeFunction() >
+            xAzureIoTProvisioningClientHandle->_internal.ullProvisioningRequestTimeout )
+        {
+            xAzureIoTProvisioningClientHandle->_internal.ullProvisioningRequestTimeout = 0;
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningSuccess );
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * Trigger state machine action base on the state.
+ *
+ **/
+static void prvTriggerAction( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    switch( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState )
+    {
+        case azureiotprovisioningWF_STATE_CONNECT:
+            prvAzureIoTProvisioningClientConnect( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_SUBSCRIBE:
+            prvAzureIoTProvisioningClientSubscribe( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_REQUEST:
+            prvAzureIoTProvisioningClientRequest( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_RESPONSE:
+            prvAzureIoTProvisioningClientParseResponse( xAzureIoTProvisioningClientHandle );
+            break;
+
+        case azureiotprovisioningWF_STATE_SUBSCRIBING:
+        case azureiotprovisioningWF_STATE_REQUESTING:
+        case azureiotprovisioningWF_STATE_COMPLETE:
+            /* None action taken here, as these states are waiting for receive path. */
+            break;
+
+        case azureiotprovisioningWF_STATE_WAITING:
+            prvAzureIoTProvisioningClientCheckTimeout( xAzureIoTProvisioningClientHandle );
+            break;
+
+        default:
+            AZLogError( ( "State not handled %d\r\n",
+                          xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+            configASSERT( false );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *  Run the workflow : trigger action on state and process receive path in MQTT loop
+ *
+ */
+static AzureIoTProvisioningClientResult_t prvAzureIoTProvisioningClientRunWorkflow( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle,
+                                                                                    uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTResult_t xResult;
+    AzureIoTProvisioningClientResult_t xReturn;
+    uint32_t ulWaitTime;
+
+    do
+    {
+        if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_COMPLETE )
+        {
+            break;
```

Comment:
We should log out `if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_COMPLETE )` case, since the way the API works & state machine someone may not realize fatal errors or even just re-running provisionion request would need a reset of handle.

Reply:
Ack.

---
In reply to: [603542867](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r603542867) [](ancestors = 603542867)